### PR TITLE
move fluentd into a daemonset

### LINKED ISSUE #23224: move fluentd into a daemonset in salt
ref #23174

### PR DESCRIPTION
**Pull request by mikedanese**
Fixes #23224 

cc @a-robinson @dchen1107

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/23306)

### CODE REVIEW

**Comment by mikedanese** [comment]
Timeout it looks like... Maybe a flake?

@k8s-bot ok to test, issue: #IGNORE

File: cluster/saltbase/salt/kube-addons/init.sls [hunk 1]
```diff
@@ -133,11 +133,11 @@ addon-dir-create:
 {% endif %}
 
 {% if pillar.get('enable_node_logging', '').lower() == 'true'
-   and pillar.get('logging_destination', '').lower() == 'elasticsearch'
+   and 'logging_destination' in pillar
    and pillar.get('enable_cluster_logging', '').lower() == 'true' %}
 /etc/kubernetes/addons/fluentd-elasticsearch:
```
Comment (a-robinson) [inline cluster/saltbase/salt/kube-addons/init.sls]: Will this always put the file under fluentd-elasticsearch? If so, does that matter?

**Comment by a-robinson** [comment]
Thanks @mikedanese. This will enable things like #8001.

Have you considered how this will interact with upgrades? Given that we upgrade masters before nodes, will upgrading to a version with this daemonset as an addon cause two fluentd instances to be running on all nodes until the nodes are upgraded?

**Comment by mikedanese** [comment]
@a-robinson good point. This might be painful...

**Comment by mikedanese** [comment]
Actually I don't think it's that bad. There are a couple ways we could do this. I'll write up the options.

**Comment by mikedanese** [comment]
The upgrade will work by restricting the daemonset to nodes with version >= 1.3. Depends on DaemonSet support https://github.com/kubernetes/kubernetes/issues/22205

**Comment by a-robinson** [comment]
That'll work beautifully if that's something we support, thanks! Let me know once this is ready to review again.

File: cluster/saltbase/salt/kube-addons/init.sls (hunk 1 above)
Comment (vishh) [inline cluster/saltbase/salt/kube-addons/init.sls]: . +1 Why not call it `/etc/kubernetes/addons/fluentd`?

**Comment by a-robinson** [comment]
Related question: was the upgrade story from 1.1 to 1.2 considered with respect to kube-proxy's move into a daemonset?

**Comment by mikedanese** [comment]
@a-robinson kube-proxy did not move into a daemonset. the upgrade from kube-proxy to static pod will occur during node upgrade.

**Comment by dnascimento** [comment]
@mikedanese can we merge it? 
It works nice. It also eases to modify the fluentd td-agent.conf using a custom image. Nice job

**Comment by a-robinson** [comment]
@mikedanese - what's the status of this? Is the required daemonset feature going to make it in?

**Comment by a-robinson** [comment]
ping

**Comment by ajohnstone** [comment]
any update on this?

**Comment by mikedanese** [comment]
@a-robinson this required some kubelet work to get into 1.3, namely figuring out an acceptable numeric version scheme to add as a label to nodes. We need this so that we can select (kubelets with verision >= version that this change goes in) with the daemonset node selector. It should be trivial to accomplish this in new deployments but upgrading old deployments is a bit trickier.

**Comment by a-robinson** [comment]
We don't have any existing standardized fluentd deployments to have to deal with upgrading. Are you saying that if we switch to deploying as a daemonset in 1.4, that we'll have a tough time upgrading when 1.5 rolls around?

**Comment by mikedanese** [comment]
@a-robinson no. If we switch to fluentd daemonset in 1.4, people will be running two fluentd pods per node until they upgrade their nodes. After the move from static pod to daemonset, upgrades should be straight forward.

**Comment by a-robinson** [comment]
I thought that was the problem originally brought up in https://github.com/kubernetes/kubernetes/pull/23306#issuecomment-199676785 that using the version selector was supposed to fix? If an old node doesn't have a version label on it, then won't it not match a label selector that requires >= a given version?

**Comment by mikedanese** [comment]
Yes, that's the issue i'm referring to. The version is not currently published as a label. It should be possible to do this with the new [NodeSelector](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1441) but that label is not currently published. We need to figure out how to map the semantic version of the kubelet binary to a float and publish that as a node label and keep it up to date. That work was never done.

**Comment by so0k** [comment]
As mentioned in the issue I linked, `/mnt/ephemeral/..` would need to be added to the fluentd volumes to support AWS.

**Comment by a-robinson** [comment]
cc @piosz @fgrzadkowski

**Comment by piosz** [comment]
I'm going to take over the effort, so closing the PR

**Comment by piosz** [comment]
@mikedanese could you please restore the branch? I'd like to cherry-pick your commit to my repo.

**Comment by mikedanese** [comment]
Done.

**Comment by piosz** [comment]
I sent a PR #32088 which supersedes this one

**Pull request closed without merging**